### PR TITLE
[VIRTS-2717] Operations API Tests

### PR DESCRIPTION
**Pull request by bleepbop**
## Description

Pytest unittests for the v2 Operations API. Note: the `DELETE` endpoint is currently untested, as there seems to be an [underlying aiohttp issue with v3.6.2](https://github.com/aio-libs/aiohttp/issues/4549), causing a failure when running with tox.

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Verified that tests work for various python versions using tox.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: tests/api/v2/handlers/test_operations_api.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    "setup_operations_api_test"
+)
+class TestOperationsApi:
+    async def test_get_operations(self, api_client, api_cookies):
+        resp = await api_client.get('/api/v2/operations', cookies=api_cookies)
+        operations_list = await resp.json()
+        assert len(operations_list) == 1
+        operation_dict = operations_list[0]
+        assert operation_dict['name'] == 'My Test Operation'
+        assert operation_dict['id'] == '123'
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:89]: Let's compare with `expected_operation` (may need to make a separate fixture or save it as a variable)

File: tests/api/v2/handlers/test_operations_api.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                       adversary={'adversary_id': '123'}, source={'id': '123'})
+        resp = await api_client.post('/api/v2/operations', json=payload)
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
+    async def test_update_operation(self, api_client, api_cookies):
+        payload = dict(state='running', obfuscator='base64')
+        resp = await api_client.patch('/api/v2/operations/123', cookies=api_cookies, json=payload)
+        assert resp.status == HTTPStatus.OK
+        op = (await BaseService.get_service('data_svc').locate('operations', {'id': '123'}))[0]
+        assert op.state == payload['state']
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:140]: we can also check to make sure other operation fields weren't changed

File: tests/api/v2/handlers/test_operations_api.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_unauthorized_update_operation(self, api_client):
+        payload = dict(state='running', obfuscator='base64')
+        resp = await api_client.patch('/api/v2/operations/123', json=payload)
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
+    async def test_get_links(self, api_client, api_cookies):
+        resp = await api_client.get('/api/v2/operations/123/links', cookies=api_cookies)
+        assert resp.status == HTTPStatus.OK
+        links = await resp.json()
+        assert len(links) == 1
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:152]: Let's also check link properties

File: tests/api/v2/handlers/test_operations_api.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_unauthorized_get_links(self, api_client):
+        resp = await api_client.get('/api/v2/operations/123/links')
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
+    async def test_get_operation_link(self, api_client, api_cookies):
+        resp = await api_client.get('/api/v2/operations/123/links/456', cookies=api_cookies)
+        assert resp.status == HTTPStatus.OK
+        link = await resp.json()
+        assert link
+        assert link['command'] == 'd2hvYW1p'
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:163]: We can compare with an expected link dictionary

File: tests/api/v2/handlers/test_operations_api.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_unauthorized_get_operations(self, api_client):
+        resp = await api_client.get('/api/v2/operations')
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
+    async def test_get_operation_by_id(self, api_client, api_cookies):
+        resp = await api_client.get('/api/v2/operations/123', cookies=api_cookies)
+        operation_dict = await resp.json()
+        assert operation_dict['name'] == 'My Test Operation'
+        assert operation_dict['id'] == '123'
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:100]: we should also test getting a non-existent operation

File: tests/api/v2/handlers/test_operations_api.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            BaseService.get_service('data_svc').ram['operations'][0].objective = test_objective
+            resp = await api_client.get('/api/v2/operations/123/report', cookies=api_cookies)
+            report = await resp.json()
+            assert report['name'] == 'My Test Operation'
+            assert report['jitter'] == '2/8'
+
+    async def test_unauthorized_get_operation_report(self, api_client):
+        resp = await api_client.get('/api/v2/operations/report')
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:118]: let's also test getting an operation report for a nonexistent operation

File: tests/api/v2/handlers/test_operations_api.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_create_operation(self, api_client, api_cookies):
+        payload = dict(name='post_test', planner={'id': '123'},
+                       adversary={'adversary_id': '123'}, source={'id': '123'})
+        resp = await api_client.post('/api/v2/operations', cookies=api_cookies, json=payload)
+        assert resp.status == HTTPStatus.OK
+        op_data = await resp.json()
+        assert op_data.get('name') == "post_test"
+        op_exists = await BaseService.get_service('data_svc').locate('operations', {'name': 'post_test'})
+        assert op_exists
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:128]: we can also test trying to create an operation using an ID that's already in use

File: tests/api/v2/handlers/test_operations_api.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
+    async def test_update_operation(self, api_client, api_cookies):
+        payload = dict(state='running', obfuscator='base64')
+        resp = await api_client.patch('/api/v2/operations/123', cookies=api_cookies, json=payload)
+        assert resp.status == HTTPStatus.OK
+        op = (await BaseService.get_service('data_svc').locate('operations', {'id': '123'}))[0]
+        assert op.state == payload['state']
+        assert op.obfuscator == payload['obfuscator']
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:142]: we should also test:
1) updating a nonexistent operation
2) trying to update fields that can't be updated

File: tests/api/v2/handlers/test_operations_api.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_get_links(self, api_client, api_cookies):
+        resp = await api_client.get('/api/v2/operations/123/links', cookies=api_cookies)
+        assert resp.status == HTTPStatus.OK
+        links = await resp.json()
+        assert len(links) == 1
+
+    async def test_unauthorized_get_links(self, api_client):
+        resp = await api_client.get('/api/v2/operations/123/links')
+        assert resp.status == HTTPStatus.UNAUTHORIZED
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:157]: also test getting links for a nonexistent op

File: tests/api/v2/handlers/test_operations_api.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_get_operation_link(self, api_client, api_cookies):
+        resp = await api_client.get('/api/v2/operations/123/links/456', cookies=api_cookies)
+        assert resp.status == HTTPStatus.OK
+        link = await resp.json()
+        assert link
+        assert link['command'] == 'd2hvYW1p'
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:164]: also test getting a nonexistent link

File: tests/api/v2/handlers/test_operations_api.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        BaseService.get_service('rest_svc').build_potential_abilities = mocker.Mock()
+        BaseService.get_service('rest_svc').build_potential_abilities.return_value = async_return([])
+        test_link = Link(command='whoami', paw='123456', id='789')
+        BaseService.get_service('rest_svc').build_potential_links = mocker.Mock()
+        BaseService.get_service('rest_svc').build_potential_links.return_value = async_return([test_link])
+        resp = await api_client.get('/api/v2/operations/123/potential-links', cookies=api_cookies)
+        result = await resp.json()
+        assert len(result) == 1
+        assert result[0]['id'] == '789'
+
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:191]: also test getting potential links for a nonexistent operation

File: tests/api/v2/handlers/test_operations_api.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    async def test_get_potential_links_by_paw(self, api_client, api_cookies, mocker, async_return, ability, executor):
+        BaseService.get_service('rest_svc').build_potential_abilities = mocker.Mock()
+        BaseService.get_service('rest_svc').build_potential_abilities.return_value = async_return([])
+        test_link = Link(command='whoami', paw='123456', id='789')
+        BaseService.get_service('rest_svc').build_potential_links = mocker.Mock()
+        BaseService.get_service('rest_svc').build_potential_links.return_value = async_return([test_link])
+        resp = await api_client.get('/api/v2/operations/123/potential-links/123', cookies=api_cookies)
+        result = await resp.json()
+        assert len(result) == 1
+        assert result[0]['id'] == '789'
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:205]: also test getting links for a paw that doesn't exist in the operation

**Review by uruwhy — CHANGES_REQUESTED** [review]
Looking great! Couple other addition's that would help with completeness:
- Testing behavior for nonexistent operation IDs / link IDs
- Testing certain edge cases, like trying to update or create a finished operation.

File: tests/conftest.py [hunk 13]
```diff
@@ -243,3 +255,74 @@ def _agent_profile(paw=None, group='red', platform='linux', executors=None, priv
         )
 
     return _agent_profile
+
+
+@pytest.fixture
+def api_client(loop, aiohttp_client):
```
Comment (uruwhy) [inline tests/conftest.py:261]: if this api client is only being used for API v2, maybe we can rename it to `api_v2_client` so it doesn't get mixed up with other api clients in other tests

File: tests/api/v2/handlers/test_operations_api.py [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+@pytest.fixture
+def test_source(loop):
+    test_fact = Fact(trait='remote.host.fqdn', value='dc')
+    test_source = Source(id='123', name='test', facts=[test_fact], adjustments=[])
+    loop.run_until_complete(BaseService.get_service('data_svc').store(test_source))
+    return test_source
+
+
+@pytest.fixture
+def expected_operation(test_adversary, test_planner, test_source):
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:55]: `test_operation` would be a better name for this fixture

File: tests/api/v2/handlers/test_operations_api.py [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+    loop.run_until_complete(BaseService.get_service('data_svc').store(finished_operation))
+
+
+@pytest.fixture
+def setup_operations_api_test(loop, api_client, expected_operation, test_agent, test_ability,
+                              active_link, finished_link):
+    test_operation = OperationSchema().load(expected_operation)
+    test_operation.agents.append(test_agent)
+    test_operation.set_start_details()
+    test_operation.adversary.atomic_ordering.append(test_ability)
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:148]: is this necessary since it's supposed to be an ad-hoc adversary?

File: tests/api/v2/handlers/test_operations_api.py [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+        resp = await api_client.get('/api/v2/operations/999/potential-links', cookies=api_cookies)
+        assert resp.status == HTTPStatus.NOT_FOUND
+
+    async def test_get_potential_links_by_paw(self, api_client, api_cookies, mocker, async_return, ability, executor):
+        BaseService.get_service('rest_svc').build_potential_abilities = mocker.Mock()
+        BaseService.get_service('rest_svc').build_potential_abilities.return_value = async_return([])
+        expected_link = Link(command='whoami', paw='123456', id='789')
+        BaseService.get_service('rest_svc').build_potential_links = mocker.Mock()
+        BaseService.get_service('rest_svc').build_potential_links.return_value = async_return([expected_link])
+        resp = await api_client.get('/api/v2/operations/123/potential-links/123', cookies=api_cookies)
```
Comment (uruwhy) [inline tests/api/v2/handlers/test_operations_api.py:375]: Maybe I'm confused, but how is this test passing if `123` is the requested paw, but the link has a paw of `123456`?

Comment (uruwhy) [reply tests/api/v2/handlers/test_operations_api.py:375]: Ah, looks like the mocked method will always return the expected link regardless of the requested paw

Comment (uruwhy) [reply tests/api/v2/handlers/test_operations_api.py:375]: We're not really testing the rest service methods here, but for consistency let's make the Link paw the same as the agent paw (123)

**Review by uruwhy — CHANGES_REQUESTED** [review]
Few small changes here and there, but overall looks great!

File: tests/api/v2/handlers/test_operations_api.py (hunk 15 above)
Comment (bleepbop) [reply tests/api/v2/handlers/test_operations_api.py:148]: That's a good point, I'll remove that

**Review by uruwhy — APPROVED** [review]
Looks good to me!

**Review by wbooth — APPROVED** [review]
